File: lib/domain/cultivation_plan/gateways/plan_copy_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      # 年度私有コピー用の狭い永続化ポート（CRUD のみ）。
      class PlanCopyGateway
        # @return [Domain::CultivationPlan::Entities::CultivationPlanEntity]
        # @raise [Domain::Shared::Exceptions::RecordNotFound]
        def find_plan(source_plan_id:)
          raise NotImplementedError
        end

        # @return [Array<Domain::CultivationPlan::Dtos::PlanCopyFieldRow>]
        def list_fields(source_plan_id:)
          raise NotImplementedError
        end

        # @return [Array<Domain::CultivationPlan::Dtos::PlanCopyCropRow>]
        def list_crops(source_plan_id:)
          raise NotImplementedError
        end

        # @return [Array<Domain::CultivationPlan::Dtos::PlanCopyFieldCultivationRow>]
        def list_field_cultivations(source_plan_id:)
          raise NotImplementedError
        end

        # @param attrs [Domain::CultivationPlan::Dtos::PlanCopyCreateAttrs]
        # @return [Domain::CultivationPlan::Entities::CultivationPlanEntity]
        def create_plan(attrs:)
          raise NotImplementedError
        end

        # @return [Domain::CultivationPlan::Dtos::PlanCopyFieldRow]
        def create_field(plan_id:, name:, area:, daily_fixed_cost:, description: nil)
          raise NotImplementedError
        end

        # @return [Domain::CultivationPlan::Dtos::PlanCopyCropRow]
        def create_crop(plan_id:, crop_id:, name:, variety:, area_per_unit:, revenue_per_area:)
          raise NotImplementedError
        end

        def create_field_cultivation(plan_id:, cultivation_plan_field_id:, cultivation_plan_crop_id:, area:, status:)
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_copy_gateway.rs
//! Ruby: plan copy persistence port.

use crate::cultivation_plan::dtos::{
    PlanCopyCreateAttrs, PlanCopyCropSnapshot, PlanCopyFieldCultivationSnapshot,
    PlanCopyFieldSnapshot, PlanCopySourcePlan,
};
use crate::cultivation_plan::entities::CultivationPlanEntity;

pub trait PlanCopyGateway: Send + Sync {
    fn find_plan(
        &self,
        source_plan_id: i64,
    ) -> Result<PlanCopySourcePlan, Box<dyn std::error::Error + Send + Sync>>;

    fn create_plan(
        &self,
        attrs: &PlanCopyCreateAttrs,
    ) -> Result<CultivationPlanEntity, Box<dyn std::error::Error + Send + Sync>>;

    fn list_fields(
        &self,
        source_plan_id: i64,
    ) -> Result<Vec<PlanCopyFieldSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn create_field(
        &self,
        plan_id: i64,
        name: &str,
        area: f64,
        daily_fixed_cost: f64,
        description: Option<&str>,
    ) -> Result<PlanCopyFieldSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn list_crops(
        &self,
        source_plan_id: i64,
    ) -> Result<Vec<PlanCopyCropSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn create_crop(
        &self,
        plan_id: i64,
        crop_id: i64,
        name: &str,
        variety: Option<&str>,
        area_per_unit: f64,
        revenue_per_area: f64,
    ) -> Result<PlanCopyCropSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn list_field_cultivations(
        &self,
        source_plan_id: i64,
    ) -> Result<Vec<PlanCopyFieldCultivationSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn create_field_cultivation(
        &self,
        plan_id: i64,
        cultivation_plan_field_id: i64,
        cultivation_plan_crop_id: i64,
        area: f64,
        status: &str,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
}
